lib/collection/src/shards/shard_holder.rs

```diff
@@ -19,41 +19,97 @@ use crate::save_on_disk::SaveOnDisk;
 use crate::shards::channel_service::ChannelService;
 use crate::shards::local_shard::LocalShard;
 use crate::shards::replica_set::{ChangePeerState, ReplicaState, ShardReplicaSet}; // TODO rename ReplicaShard to ReplicaSetShard
-use crate::shards::shard::{PeerId, ShardId};
+use crate::shards::shard::{PeerId, ShardId, ShardKey};
 use crate::shards::shard_config::{ShardConfig, ShardType};
 use crate::shards::shard_versioning::latest_shard_paths;
 use crate::shards::transfer::shard_transfer::{ShardTransfer, ShardTransferKey};
 use crate::shards::CollectionId;
 
 const SHARD_TRANSFERS_FILE: &str = "shard_transfers";
+const SHARD_KEY_MAPPING_FILE: &str = "shard_key_mapping.json";
 
 pub struct ShardHolder {
     shards: HashMap<ShardId, ShardReplicaSet>,
     pub(crate) shard_transfers: SaveOnDisk<HashSet<ShardTransfer>>,
     ring: HashRing<ShardId>,
+    key_mapping: SaveOnDisk<HashMap<ShardKey, HashSet<ShardId>>>,
 }
 
 pub type LockedShardHolder = RwLock<ShardHolder>;
 
 impl ShardHolder {
     pub fn new(collection_path: &Path, hashring: HashRing<ShardId>) -> CollectionResult<Self> {
         let shard_transfers = SaveOnDisk::load_or_init(collection_path.join(SHARD_TRANSFERS_FILE))?;
+        let key_mapping = SaveOnDisk::load_or_init(collection_path.join(SHARD_KEY_MAPPING_FILE))?;
         Ok(Self {
             shards: HashMap::new(),
             shard_transfers,
             ring: hashring,
+            key_mapping,
         })
     }
 
-    pub fn add_shard(&mut self, shard_id: ShardId, shard: ShardReplicaSet) {
+    pub fn get_shard_id_to_key_mapping(&self) -> HashMap<ShardId, ShardKey> {
+        self.key_mapping
+            .read()
+            .iter()
+            .flat_map(|(key, shard_ids)| {
+                shard_ids
+                    .iter()
+                    .map(move |shard_id| (*shard_id, key.clone()))
+            })
+            .collect()
+    }
```

Comment:
This is not very efficient. I'm not sure how hot this is yet, but if it is, we might want to create a mapping in both directions.